### PR TITLE
feat/core-base-strategy

### LINKED ISSUE #591: Ensure default receive is virtual and can be called only by Allo

### PR DESCRIPTION
**Pull request by 0xOneTony**
Closes #591 , #590 

Most Basic Strategy contract that should be used for creating extensions or more complex strategies.

⚠️ It's important to note that I didn't remove the current `BaseStrategy` yet, as it would break all the strategies

### CODE REVIEW

File: contracts/core/interfaces/IBaseStrategy.sol [hunk 1]
```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.19;
```
Comment (0xAustrian) [inline contracts/core/interfaces/IBaseStrategy.sol:2]: Remember to use fixed solc version.

**Review by 0xAustrian — APPROVED** [review]

**Review by ilpepepig — APPROVED** [review]

File: contracts/core/interfaces/IBaseStrategy.sol [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    /// @return _poolId The ID of the pool
+    function getPoolId() external view returns (uint256 _poolId);
+
+    /// @notice Checks the amount of tokens in the pool.
+    /// @return _poolAmount The balance of the pool
+    function getPoolAmount() external view returns (uint256 _poolAmount);
+
+    /// @notice Gets the timestamp of the phase at the given index.
+    /// @param _index The index of the phase
+    function getTimestampPhase(uint256 _index) external view returns (uint256 _timestamp);
```
Comment (thelostone-mc) [inline contracts/core/interfaces/IBaseStrategy.sol:72]: What's this about ?

**Comment by owocki** [comment]
you love to see it :)

what are we thinking the taxonomy of strategies should be

is it
```
basestrategy
strategy strategy strategy strategy strategy strategy 
```

or will there be multiple layers like

```
basestrategy
quadratic_straegy                              retro_strategy                        xyz_strategy
vanillaQAstrategy MACIQFstrategy            vanilla_retro                          strawberry_xyz_strategy  
```

**Comment by ilpepepig** [comment]
Thinking again about this, I'm not sure about the use modifiers. I think their current logic belong in the `_beforeX()` functions and we should give custom strategy devs more flexibility.

For example, why is it assumed that `withdraw()` should be called by the Pool Manager? `_checkOnlyPoolManager` could be called from `_beforeWithdraw` if needed, and maybe by default. Also, why is the `onlyAllo` modifier not used here?

In addition, if custom strategies are not whitelisted, enforcing `onlyAllo` is not really possible but should be convinient. Can we simply remove this requirement? What would be the consequences?

Regarding `onlyInitialized`, currently strategies are cloned by Allo and initialized in the same call, which means that all strategies are by default initialized. Can we remove this modifier then? If a custom strategy requires a different initialization logic, checking that it happened could be added to `_beforeX()`.

**Comment by 0xOneTony** [comment]
> what are we thinking the taxonomy of strategies should be

@owocki So we should be providing 2 things to devs
1. tools to build easier their custom strategy (making use of base strategy, libraries, extensions)
2. use one of our example strategies (QVStrategy, QFStrategy, RetroPGFStrategy, RFPStrategy ....)

so the taxonomy should be 

```
BaseStrategy    /libraries   /extensions
/QuadraticVotingStrategies       /RFPStrategies      /DonationVotingStrategies
/RetroStrategies         /LTIPStrategies        /MACIStrategies
```

**Review by thelostone-mc — APPROVED** [review]
All look good except the reason/need for getTimestampPhase

File: contracts/core/interfaces/IBaseStrategy.sol (hunk 2 above)
Comment (0xOneTony) [reply contracts/core/interfaces/IBaseStrategy.sol:72]: yea not really needed, now that i think about it.

**Review by 0xAustrian — APPROVED** [review]

**Review by ilpepepig — APPROVED** [review]

**Pull request merged**